lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,374 @@
+        }
+
+
+        /**
+         * Opens a connection to a MySQL database.
+         *
+         * @param {Function} callback
+         */
+        connect(callback) {
+            var connectionOptions, mysqlOptions;
+
+            /*
+            These are not all of the options that a mysql connection can take,
+            only what is most relevant to this plugin.
+            See the MySQL documentation for the rest.
+            */
+            mysqlOptions = [
+                "database",
+                "host",
+                "password",
+                "port",
+                "table",
+                "user"
+            ];
+            connectionOptions = {};
+
+            mysqlOptions.forEach((key) => {
+                if (this.config[key]) {
+                    connectionOptions[key] = this.config[key];
+                }
+            });
+
+            this.connection = mysql.createConnection(connectionOptions);
+            this.connection.connect(callback);
+        }
+
+
+        /**
+         * Severs the connection to the database.
+         *
+         * @param {Function} callback
+         */
+        disconnect(callback) {
+            this.connection.end(callback);
+        }
+
+
+        /**
+         * Starts polling and restoring.
+         *
+         * @param {Function} callback
+         */
+        listen() {
+            this.currentlyRestoring = true;
+            this.currentlyPolling = true;
+            restore(this);
+            poll(this);
+        }
+
+
+        /**
+         * Resets the relevant flags and clears the timers, which stops both
+         * polling and restoring.
+         */
+        pausePolling() {
+            if (this.restorer) {
+                timers.clearTimeout(this.restorer);
+                this.restorer = null;
+                this.currentlyRestoring = false;
+            }
+
+            if (this.poller) {
+                timers.clearTimeout(this.poller);
+                this.poller = null;
+                this.currentlyPolling = false;
+            }
+        }
+
+
+        /**
+         * Sets the relevant flags and starts polling and restoring.
+         *
+         * @param {Function} callback
+         */
+        resumePolling() {
+            this.currentlyPolling = true;
+            this.currentlyRestoring = true;
+            poll(this);
+            restore(this);
+        }
+
+
+        /**
+         * Sets up the parameters for trySendMessage and writeRecord and calls
+         * the former.
+         *
+         * @param {Object} message
+         * @param {string} topics
+         * @param {Function} callback
+         */
+        sendMessage(message, topics, callback) {
+            var hash, params;
+
+            cycleCounter = 0;
+            hash = crypto.createHash("sha256").update(message).digest("hex");
+
+            params = {
+                hash,
+                message,
+                topics,
+                callback
```

Comment:
Would you pull this out of the params?  It doesn't match the message we are trying to put *in* and is instead an *out* of the program.